share/spack/docker/entrypoint.bash

```diff
@@ -1,45 +1,143 @@
-#! /usr/bin/env bash -e
+#! /usr/bin/env bash
 #
 # Copyright 2013-2020 Lawrence Livermore National Security, LLC and other
 # Spack Project Developers. See the top-level COPYRIGHT file for details.
 #
 # SPDX-License-Identifier: (Apache-2.0 OR MIT)
 
-unset CURRENTLY_BUILDING_DOCKER_IMAGE
+mode=oneshot
 
-if [ "$1" '=' 'docker-shell' ] ; then
-    if [ -t 0 ] ; then
-        exec bash -il
-    else
-        (
-            echo -n "It looks like you're trying to run an intractive shell"
-            echo -n " session, but either no psuedo-TTY is allocateed for this"
-            echo -n " container's STDIN, or it is closed."
-            echo
+if [ "$( basename "$0" )" '=' 'spack-env' ] ; then
+    mode=spackenv
+elif [ "$( basename "$0" )" '=' 'docker-shell' ] ; then
+    mode=dockershell
+elif [ "$( basename "$0" )" '=' 'interactive-shell' ] ; then
+    mode=interactiveshell
+elif [ "$1" '=' 'docker-shell' ] ; then
+    mode=dockershell
+    shift
+elif [ "$1" '=' 'interactive-shell' ] ; then
+    mode=interactiveshell
+    shift
+fi
 
-            echo -n "Make sure you run docker with the --interactive and --tty"
-            echo -n " options."
-            echo
-        ) >&2
+case "$mode" in
+    "spackenv")
+        # Scenario 1: Run as if the image had no ENTRYPOINT
```

Comment:
Thanks for these descriptions of how the various entrypoints are designed to work.